distmv/studentst.go

```diff
@@ -0,0 +1,335 @@
+	dim        int
+}
+
+// NewStudentsT creates a new StudentsT with the given scale, mean, and covariance
+// parameters.
+//
+// NewStudentsT panics if len(mu) == 0, or if len(mu) != sigma.Symmetric(). If
+// the covariance matrix is not positive-definite, nil is returned and the
+// boolean is false.
+func NewStudentsT(nu float64, mu []float64, sigma mat64.Symmetric, src *rand.Rand) (*StudentsT, bool) {
+	if len(mu) == 0 {
+		panic(badZeroDimension)
+	}
+	dim := sigma.Symmetric()
+	if dim != len(mu) {
+		panic(badSizeMismatch)
+	}
+
+	s := &StudentsT{
+		nu:  nu,
+		mu:  make([]float64, dim),
+		dim: dim,
+		src: src,
+	}
+	copy(s.mu, mu)
+
+	ok := s.chol.Factorize(sigma)
+	if !ok {
+		return nil, false
+	}
+	s.lower.LFromCholesky(&s.chol)
+	s.logSqrtDet = 0.5 * s.chol.LogDet()
+	return s, true
+}
+
+// ConditionStudentsT returns the Student's T distribution that is the receiver
+// conditioned on the input evidence. The returned Student's T has dimension
+// n - len(observed), where n is the dimension of the original receiver.
+// The dimension order is preserved during conditioning, so if the value
+// of dimension 1 is observed, the returned normal represents dimensions {0, 2, ...}
+// of the original Student's T distribution.
+//
+// ConditionStudentsT returns {nil, false} if there is a failure during the update.
+// Mathematically this is impossible, but can occur with finite precision arithmetic.
+func (s *StudentsT) ConditionStudentsT(observed []int, values []float64, src *rand.Rand) (*StudentsT, bool) {
+	if len(observed) == 0 {
+		panic("studentst: no observed value")
+	}
+	if len(observed) != len(values) {
+		panic(badInputLength)
+	}
+
+	for _, v := range observed {
+		if v < 0 || v >= s.dim {
+			panic("studentst: observed value out of bounds")
+		}
+	}
+
+	s.setSigma()
+
+	newNu, newPhi, newK := studentsTConditional(observed, values, s.nu, s.mu, s.sigma)
+	if newPhi == nil {
+		return nil, false
+	}
+
+	return NewStudentsT(newNu, newPhi, newK, src)
+
+}
+
+// findUnob returns the unobserved variables (the complementary set to observed).
+// findUnob panics if any value repeated in observed.
+func findUnob(observed []int, dim int) (unobserved []int) {
+	ob := len(observed)
+	unob := dim - ob
+	obMap := make(map[int]struct{})
+	for _, v := range observed {
+		if _, ok := obMap[v]; ok {
+			panic("distmv: observed dimension occurs twice")
+		}
+		obMap[v] = struct{}{}
+	}
+	unobserved = make([]int, 0, unob)
+	for i := 0; i < dim; i++ {
+		if _, ok := obMap[i]; !ok {
+			unobserved = append(unobserved, i)
+		}
+	}
+	return unobserved
+}
+
+// studentsTConditional updates a Students T distribution based on the observed samples
+// (see documentation for the public function). The Gaussian conditional update
+// is treated as a special case when  nu == math.Inf(1).
+func studentsTConditional(observed []int, values []float64, nu float64, mu []float64, sigma mat64.Symmetric) (newNu float64, newMean []float64, newSigma mat64.Symmetric) {
```

Comment:
This can return a concrete `*mat64.SymDense`, though I'm not fussed about this either way.